haddock-api/src/Haddock/Convert.hs

```diff
@@ -610,3 +610,32 @@ synifyFamInst fi opaque = do
     ts' = synifyTypes ts
     annot_ts = zipWith3 annotHsType is_poly_tvs ts ts'
     is_poly_tvs = mkIsPolyTvs (tyConVisibleTyVars fam_tc)
+
+-- | A version of 'TcType.tcSplitSigmaTy' that preserves type synonyms.
+--
+-- See https://github.com/haskell/haddock/issues/879 for why we need this.
```

Comment:
Perhaps it would be worth expanding this into a Note, and referencing it from `tcSplitSigmaTyPreserveSynonyms`, `tcSplitForAllTysPreserveSynonyms`, and `tcSplitPhiTyPreserveSynonyms`. In general, I think there should be an invariant that no functions in `Convert` should ever expand type synonyms (since `Convert` produces entities that are user-facing)—it would be worth mentioning this invariant in the Note.

Reply:
Done. Can you check the note I added?